### PR TITLE
Variable and type declarations

### PR DESCRIPTION
**Pull request by m-lyons**
Finally got this done. This doesn't include support for port types and node value types because they are done in a weird way. Fixing them is going to be something for the co-op term.

### CODE REVIEW

**Comment by psocha** [comment]
Do you see any room for adding new unit tests for this functionality in `TestExpressionContextVisitor`?

File: src/main/antlr4/Manifold.g4 [hunk 1]
```diff
@@ -57,11 +57,13 @@ typevalue:
   | functionTypeValue # FunctionType
   ;
 
-// TODO: implement type declarations and undefined variable declarations
-//declaration:
-//    type namespacedIdentifier # UndefinedDeclaration
-//  | TYPE_KEYWORD namespacedIdentifier '=' type # TypeDeclaration
-//  ;
+undefinedDeclaration: type namespacedIdentifier;
```
Comment (nikklassen) [inline src/main/antlr4/Manifold.g4]: should probably call this undefined**Type**Declaration

File: src/main/java/org/manifold/compiler/front/ExpressionContextVisitor.java [hunk 2]
```diff
@@ -300,11 +297,27 @@ public ExpressionVertex visitNamespacedIdentifier(
...
       // construct a constant value vertex with the identifier's value
       ConstantValueVertex vReserved = new ConstantValueVertex(exprGraph,
           ReservedIdentifiers.getInstance().getValue(id));
       exprGraph.addVertex(vReserved);
-      return vReserved;
+
+      // create a variable vertex to represent this constant in the graph,
+      // this prevents unnecessary copies of vertices.
+      VariableReferenceVertex v = new VariableReferenceVertex(exprGraph, id);
+      try {
```
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/ExpressionContextVisitor.java]: This try/catch variable creation seems pretty common in this file. Can we just pull variable creation out into a function/generalize it?

File: src/main/java/org/manifold/compiler/front/FunctionValueVertex.java [hunk 3]
```diff
@@ -67,24 +67,29 @@ public void elaborate() throws Exception {
     TupleTypeValue outputType = (TupleTypeValue) type.getOutputType();
     MappedArray<String, ExpressionEdge> inputEdges = new MappedArray<>();
     MappedArray<String, ExpressionEdge> outputEdges = new MappedArray<>();
-    for (MappedArray<String, TypeValue>.Entry argName : inputType.getSubtypes()) {
+    inputType.getSubtypes().forEach((argName) -> {
       ExpressionEdge e = new ExpressionEdge(null, null); // I know what I'm doing
```
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/FunctionValueVertex.java]: Do you though?

File: src/main/java/org/manifold/compiler/front/FunctionValueVertex.java [hunk 4]
```diff
@@ -67,24 +67,29 @@ public void elaborate() throws Exception {
...
-    for (MappedArray<String, TypeValue>.Entry typeEntry : outputType.getSubtypes()) {
+    });
+    outputType.getSubtypes().forEach((typeEntry) -> {
       String argName = typeEntry.getKey();
       // "name resolution": look for a variable reference vertex with this name in the subgraph
       for (Map.Entry<VariableIdentifier, VariableReferenceVertex> vRef
           : functionBody.getVariableVertices().entrySet()) {
+        if (vRef.getKey() == null) {
+          // then the user tried to have an output tuple without named keys
+          throw new FrontendBuildException("The function output type is an anonymous tuple." +
```
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/FunctionValueVertex.java]: Probably want a newline/whitespace and the end of this

File: src/main/java/org/manifold/compiler/front/VariableReferenceVertex.java [hunk 5]
```diff
@@ -106,8 +106,16 @@ public boolean isRuntimeKnowable() {
 
   @Override
   public void elaborate() throws Exception {
-    ExpressionEdge e = findAssigningEdge();
-    e.getSource().elaborate();
+    ExpressionEdge assigningEdge = findAssigningEdge();
+    ExpressionEdge e = null;
+    // continue to elaborate the assigning edge until it stays
+    // in the graph upon elaboration because vertex elaboration
+    // can lead to the edge being replaced in the graph
```
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/VariableReferenceVertex.java]: What was the consequence of this?

File: src/main/java/org/manifold/compiler/front/VariableDeclarationVertex.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  @Override
+  public void elaborate() throws Exception {
+    super.elaborate();
+    type = super.getType();
+    vType.elaborate();
+    Value value = vType.getValue();
+    TypeValue declaredType = TypeAssertions.assertIsType(value);
+    // check if assigned value is of the correct declaredType
+    if (!declaredType.isSubtypeOf(type)) {
```
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/VariableDeclarationVertex.java]: Wouldn't it make more sense to put this part in the `verify` function?

**Comment by nikklassen** [comment]
Cool

File: src/main/java/org/manifold/compiler/front/VariableReferenceVertex.java (hunk 5 above)
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/VariableReferenceVertex.java]: Variables end up still pointing to the old unelaborated vertex. It's bad.

File: src/main/java/org/manifold/compiler/front/VariableDeclarationVertex.java (hunk 6 above)
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/VariableDeclarationVertex.java]: Nah

File: src/main/java/org/manifold/compiler/front/ExpressionContextVisitor.java (hunk 2 above)
Comment (nikklassen) [inline src/main/java/org/manifold/compiler/front/ExpressionContextVisitor.java]: Nope, different constructors are required. Not worth trying to extract it.

**Pull request merged**